samples/triangle/main.c

```diff
@@ -0,0 +1,281 @@
+/*
+ * This sample provides a very basic demonstration of 3D rendering on the Xbox,
+ * using pbkit. Based on the pbkit demo sources.
+ */
+#include <hal/video.h>
+#include <hal/xbox.h>
+#include <math.h>
+#include <pbkit/pbkit.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strings.h>
+#include <xboxkrnl/xboxkrnl.h>
+#include <xboxrt/debug.h>
+
+/* Restart after a number of milliseconds (0 for never) */
+const int reboot_after = 10*1000;
+
+static uint32_t *alloc_vertices;
+static uint32_t  num_vertices;
+static float     m_viewport[4][4];
+
+typedef struct {
+    float pos[3];
+    float color[3];
+} __attribute__((packed)) ColoredVertex;
+
+static const ColoredVertex verts[] = {
+    //  X     Y     Z       R     G     B
+    {{-1.0, -1.0,  1.0}, { 0.1,  0.1,  0.6}},
+    {{-1.0,  1.0,  1.0}, { 0.0,  0.0,  0.0}},
+    {{ 1.0,  1.0,  1.0}, { 0.0,  0.0,  0.0}},
+    {{-1.0, -1.0,  1.0}, { 0.1,  0.1,  0.6}},
+    {{ 1.0,  1.0,  1.0}, { 0.0,  0.0,  0.0}},
+    {{ 1.0, -1.0,  1.0}, { 0.1,  0.1,  0.6}},
+    {{-1.0, -1.0,  1.0}, { 1.0,  0.0,  0.0}},
+    {{ 0.0,  1.0,  1.0}, { 0.0,  1.0,  0.0}},
+    {{ 1.0, -1.0,  1.0}, { 0.0,  0.0,  1.0}},
+};
+
+#define MASK(mask, val) (((val) << (ffs(mask)-1)) & (mask))
+
+static void matrix_viewport(float out[4][4], float x, float y, float width, float height, float z_min, float z_max);
+static void init_shader(void);
+static void set_attrib_pointer(unsigned int index, unsigned int format, unsigned int size, unsigned int stride, const void* data);
+static void draw_arrays(unsigned int mode, int start, int count);
+
+/* Main program function */
+void main(void)
+{
+    uint32_t *p;
+    int       i, status;
+    int       width, height;
+    int       start, last, now;
+    int       fps, frames, frames_total;
+
+    if ((status = pb_init())) {
+        debugPrint("pb_init Error %d\n", status);
+        XSleep(2000);
+        XReboot();
+        return;
+    }
+
+    pb_show_front_screen();
+
+    /* Basic setup */
+    width = pb_back_buffer_width();
+    height = pb_back_buffer_height();
+
+    /* Load constant rendering things (shaders, geometry) */
+    init_shader();
+    alloc_vertices = MmAllocateContiguousMemoryEx(sizeof(verts), 0, 0x3ffb000, 0, 0x404);
```

Comment:
0x404 and 0x3ffb000; I assume one is the permissions or flags and the other one is a min/max address?